ddl/partition.go

```diff
@@ -1911,27 +1911,157 @@ func onTruncateTablePartition(d *ddlCtx, t *meta.Meta, job *model.Job) (int64, e
+		job.SchemaState = model.StateDeleteReorganization
+		ver, err = updateVersionAndTableInfo(d, t, job, tblInfo, true)
+	case model.StateDeleteReorganization:
+		// Step1: clear global index rows.
+		physicalTableIDs := oldIDs
+		oldTableInfo := getTableInfoWithOriginalPartitions(tblInfo)
+		tbl, err := getTable(d.store, job.SchemaID, oldTableInfo)
+		if err != nil {
+			return ver, errors.Trace(err)
+		}
+		dbInfo, err := t.GetDatabase(job.SchemaID)
+		if err != nil {
+			return ver, errors.Trace(err)
+		}
+		// If table has global indexes, we need reorg to clean up them.
+		if pt, ok := tbl.(table.PartitionedTable); ok && hasGlobalIndex(tblInfo) {
+			// Build elements for compatible with modify column type. elements will not be used when reorganizing.
+			elements := make([]*meta.Element, 0, len(tblInfo.Indices))
+			for _, idxInfo := range tblInfo.Indices {
+				if idxInfo.Global {
+					elements = append(elements, &meta.Element{ID: idxInfo.ID, TypeKey: meta.IndexElementKey})
+				}
+			}
+			sctx, err1 := w.sessPool.Get()
+			if err1 != nil {
+				return ver, err1
+			}
+			defer w.sessPool.Put(sctx)
+			rh := newReorgHandler(sess.NewSession(sctx))
+			reorgInfo, err := getReorgInfoFromPartitions(d.jobContext(job.ID), d, rh, job, dbInfo, pt, physicalTableIDs, elements)
+
+			if err != nil || reorgInfo.first {
+				// If we run reorg firstly, we should update the job snapshot version
+				// and then run the reorg next time.
+				return ver, errors.Trace(err)
+			}
+			err = w.runReorgJob(rh, reorgInfo, tbl.Meta(), d.lease, func() (dropIndexErr error) {
+				defer tidbutil.Recover(metrics.LabelDDL, "onDropTablePartition",
+					func() {
+						dropIndexErr = dbterror.ErrCancelledDDLJob.GenWithStack("drop partition panic")
+					}, false)
+				return w.cleanupGlobalIndexes(pt, physicalTableIDs, reorgInfo)
+			})
+			if err != nil {
+				if dbterror.ErrWaitReorgTimeout.Equal(err) {
+					// if timeout, we should return, check for the owner and re-wait job done.
+					return ver, nil
+				}
+				return ver, errors.Trace(err)
+			}
+		}
+
+		// Step2: generate new partition ids
+		newPartitions := make([]model.PartitionDefinition, 0, len(oldIDs))
+		pi.Definitions = pi.OriginalDefinitions
+		for _, oldID := range oldIDs {
+			for i := 0; i < len(pi.Definitions); i++ {
+				def := &pi.Definitions[i]
+				if def.ID == oldID {
+					pid, err1 := t.GenGlobalID()
+					if err1 != nil {
+						return ver, errors.Trace(err1)
+					}
+					def.ID = pid
+					// Shallow copy only use the def.ID in event handle.
+					newPartitions = append(newPartitions, *def)
+					break
+				}
+			}
+		}
+		if len(newPartitions) == 0 {
+			job.State = model.JobStateCancelled
+			return ver, table.ErrUnknownPartition.GenWithStackByArgs(fmt.Sprintf("pid:%v", oldIDs), tblInfo.Name.O)
+		}
+
+		if err = clearTruncatePartitionTiflashStatus(tblInfo, newPartitions, oldIDs); err != nil {
+			job.State = model.JobStateCancelled
```

Comment:
If we canceled `ActionTruncateTablePartition` job, we should consider how to rollback it.

Reply:
resolved